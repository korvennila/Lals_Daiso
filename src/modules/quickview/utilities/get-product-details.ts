/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */

/* eslint-disable no-duplicate-imports */
import { IActionContext } from '@msdyn365-commerce/core';
import { getCartStateData, GetCartStateDataInput } from '@msdyn365-commerce/global-state';
import { ProductDimension } from '@msdyn365-commerce/retail-proxy';
import {
    getDimensionsForSelectedVariant,
    GetDimensionsForSelectedVariantInput,
    getPriceForSelectedVariant,
    getProductAvailabilitiesForSelectedVariant,
    getRatingsSummary,
    GetRatingsSummaryInput,
    getRnrEndpointId,
    getRnrPartnerId,
    getSelectedVariant,
    PriceForSelectedVariantInput,
    ProductAvailabilitiesForSelectedVariantInput,
    SelectedVariantInput
} from '@msdyn365-commerce-modules/retail-actions';

import { IProductDetails } from '@msdyn365-commerce-modules/buybox';

/**
 * Calls the Retail API and returns the product details.
 * @param productId
 * @param channelId
 * @param ctx
 * @param selectedDimensions - Dimensions previously selected for the product variant.
 */
export async function getProductDetails(
    productId: number,
    channelId: number,
    ctx: IActionContext,
    selectedDimensions?: ProductDimension[]
): Promise<IProductDetails> {
    const productDetails: IProductDetails = {};

    if (productId > 0 && channelId > 0) {
        const tenantId = getRnrPartnerId(ctx);
        const serviceEndPoint = getRnrEndpointId(ctx);

        // Get Product Details
        await getSelectedVariant(new SelectedVariantInput(productId, channelId, selectedDimensions, 'get', ctx.requestContext), ctx)
            .then(async productResult => {
                if (productResult) {
                    productDetails.product = productResult;
                }
            })
            .catch(error => {
                ctx.telemetry.exception(error);
            });

        // Get Product Availability
        await getProductAvailabilitiesForSelectedVariant(new ProductAvailabilitiesForSelectedVariantInput(productId, channelId), ctx)
            .then(productAvailabilities => {
                if (productAvailabilities) {
                    productDetails.productAvailableQuantity = productAvailabilities;
                }
            })
            .catch(error => {
                ctx.telemetry.exception(error);
            });

        // Get Product Price
        await getPriceForSelectedVariant(new PriceForSelectedVariantInput(productId, channelId, undefined), ctx)
            .then(productPrice => {
                if (productPrice) {
                    productDetails.productPrice = productPrice;
                }
            })
            .catch(error => {
                ctx.telemetry.exception(error);
            });

        // Get product dimensions
        await getDimensionsForSelectedVariant(
            new GetDimensionsForSelectedVariantInput(productId, channelId, selectedDimensions, ctx.requestContext),
            ctx
        )
            .then(productDimensions => {
                if (productDimensions) {
                    productDetails.productDimensions = productDimensions;
                }
            })
            .catch(error => {
                ctx.telemetry.exception(error);
            });

        // Get Cart State
        await getCartStateData(new GetCartStateDataInput(ctx.requestContext.apiSettings, ctx.requestContext), ctx)
            .then(cartState => {
                productDetails.cart = cartState;
            })
            .catch(error => {
                ctx.telemetry.exception(error);
            });

        // Get Product Ratings
        if (tenantId && serviceEndPoint) {
            await getRatingsSummary(new GetRatingsSummaryInput(productId.toString(), tenantId, serviceEndPoint), ctx)
                .then(ratingsSummaryData => {
                    if (ratingsSummaryData) {
                        productDetails.ratingsSummary = ratingsSummaryData;
                    }
                })
                .catch(error => {
                    ctx.telemetry.exception(error);
                });
        }
    }

    return productDetails;
}
