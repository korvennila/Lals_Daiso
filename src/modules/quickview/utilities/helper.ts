/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */

/* eslint-disable no-duplicate-imports */
import { IRequestContext } from '@msdyn365-commerce/core';

/**
 * Gets correct alt text for image based on configuration of site.
 * @param context Current request context.
 * @param shouldUseCmsAltText Should use CMS alt text.
 * @param productName Product name.
 * @param imageAltText Alt-text from CMS image.
 */
export function getAltText(
    context: IRequestContext,
    shouldUseCmsAltText: boolean,
    productName?: string,
    imageAltText?: string
): string | undefined {
    if (!context?.app?.config?.OmniChannelMedia) {
        return productName;
    }
    if (shouldUseCmsAltText) {
        return imageAltText;
    }
    return productName;
}
