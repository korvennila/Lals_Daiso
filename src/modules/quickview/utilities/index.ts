/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */

export * from './get-product-details';
export * from './get-valid-product-images';
