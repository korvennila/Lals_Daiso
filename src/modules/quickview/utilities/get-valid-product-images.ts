/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */

/* eslint-disable no-duplicate-imports */
import { commerceApiRequest, getCatalogId, IActionContext, IImageData, IImageSettings } from '@msdyn365-commerce/core';
import { MediaLocation, SimpleProduct } from '@msdyn365-commerce/retail-proxy';
import { getMediaLocationsForSelectedVariant, MediaLocationsForSelectedVariantInput } from '@msdyn365-commerce-modules/retail-actions';
import { getAltText } from './helper';

export async function getValidProductImages(
    productId: number,
    channelId: number,
    actionContext: IActionContext,
    imageSettings: IImageSettings,
    selectedProduct?: SimpleProduct
): Promise<IImageData[]> {
    const catalogId = getCatalogId(actionContext.requestContext);
    const actionInput = new MediaLocationsForSelectedVariantInput(productId, channelId, selectedProduct, catalogId);

    return getMediaLocationsForSelectedVariant(actionInput, actionContext)
        .then(mediaLocations => {
            if (mediaLocations) {
                return Promise.all(mediaLocations.map(async mediaLocation => validateMediaLocaionAsync(mediaLocation, imageSettings))).then(
                    pairs => {
                        return pairs.filter(pair => pair[1]).map(pair => pair[0]);
                    }
                );
            }
            return [];
        })
        .catch(error => {
            actionContext.telemetry.exception(error);
            actionContext.telemetry.debug('Unable to get Media Locations for Selected Variant');
            return [];
        });
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-unused-vars-experimental -- .
async function validateMediaLocaionAsync(mediaLocation: MediaLocation, imageSettings?: IImageSettings): Promise<[IImageData, boolean]> {
    const imageData = {
        src: mediaLocation.Uri || '',
        altText: mediaLocation.AltText || ''
    };

    if (imageData.src === '') {
        return [imageData, false];
    }

    return new Promise<[IImageData, boolean]>(resolve => {
        try {
            const http = new XMLHttpRequest();
            http.open('HEAD', imageData.src, true);

            http.addEventListener('load', () => {
                resolve([imageData, http.status === 200 || http.status === 201]);
            });

            http.addEventListener('error', () => {
                resolve([imageData, false]);
            });

            http.send();
        } catch {
            resolve([imageData, false]);
        }
    });
}

/**
 * Gets product variant images on dimension change of buy-box module for Omni scenario.
 * @param productId Product record id.
 * @param shouldUseCmsAltText Alt-text source to use.
 * @param channelId Current channel id.
 * @param actionContext Current action context.
 * @param productName Product name.
 * @param selectedProduct Currently selected variant product.
 * @returns List of images.
 */

export async function getProductImages(
    productId: number,
    shouldUseCmsAltText: boolean,
    channelId: number,
    actionContext: IActionContext,
    productName?: string,
    selectedProduct?: SimpleProduct
): Promise<IImageData[]> {
    const catalogId = getCatalogId(actionContext.requestContext);
    if (actionContext.requestContext?.params?.isPreview) {
        const response = await commerceApiRequest(actionContext.requestContext, 'msdyn365-ochannel-product-info', 'post', {
            productId: productId,
            channelId: channelId,
            catalogId: catalogId
        });
        if (response.status !== 200) {
            actionContext.telemetry.debug('Unable to get Media Locations for Selected Variant for preview scenario');
            return [];
        }
        const previewImageData = response?.data as MediaLocation[];
        if (previewImageData) {
            return previewImageData.map(mediaLocation => {
                return {
                    src: mediaLocation.Uri ?? '',
                    altText: getAltText(actionContext.requestContext, shouldUseCmsAltText, productName, mediaLocation.AltText),
                    additionalProperties: mediaLocation.IsApplicableForChildEntities
                        ? {
                              isApplicableForChildEntities: 'true'
                          }
                        : undefined
                };
            });
        }
        return [];
    }
    const actionInput = new MediaLocationsForSelectedVariantInput(productId, channelId, selectedProduct, catalogId);
    return getMediaLocationsForSelectedVariant(actionInput, actionContext)
        .then(mediaLocations => {
            if (mediaLocations) {
                return mediaLocations.map(mediaLocation => {
                    return {
                        src: mediaLocation.Uri ?? '',
                        altText: getAltText(actionContext.requestContext, shouldUseCmsAltText, productName, mediaLocation.AltText),
                        additionalProperties: mediaLocation.IsApplicableForChildEntities
                            ? {
                                  isApplicableForChildEntities: 'true'
                              }
                            : undefined
                    };
                });
            }
            return [];
        })
        .catch(error => {
            actionContext.telemetry.exception(error);
            actionContext.telemetry.debug('Unable to get Media Locations for Selected Variant');
            return [];
        });
}
