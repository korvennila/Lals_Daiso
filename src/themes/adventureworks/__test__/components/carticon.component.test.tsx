/*!
 * Copyright (c) Microsoft Corporation.
 * All rights reserved. See LICENSE in the project root for license information.
 */

import { buildMockCoreContext, buildMockModuleProps, IComponentProps, ICoreContext } from '@msdyn365-commerce/core';
import { ICartState } from '@msdyn365-commerce/global-state';
import { Cart } from '@msdyn365-commerce/retail-proxy';
import { ITelemetryContent } from '@msdyn365-commerce-modules/utilities';
import { render } from 'enzyme';
import * as React from 'react';

import CartIcon, { ICartIconData } from '../../views/components/carticon.component';

describe('CartIcon', () => {
    const mockCoreContext: ICoreContext = buildMockCoreContext({ app: { config: { hideRating: false } } }) as ICoreContext;
    const telemetryContent = { pageName: 'Homepage', moduleName: 'Add to Cart', telemetry: mockCoreContext.telemetry } as ITelemetryContent;
    const moduleProps: IComponentProps<ICartIconData> = buildMockModuleProps({}, {}) as IComponentProps<ICartIconData>;

    // @ts-expect-error partial mock
    const subcart: Readonly<Cart> = {
        IsRequiredAmountPaid: false,
        IsReturnByReceipt: false,
        IsSuspended: false,
        LoyaltyCardId: '55182',
        MerchantProperties: [],
        Name: 'CART',
        NetPrice: 1580,
        OrderNumber: 'ORDER NUMBER',
        OtherChargeAmount: 0,
        OverriddenDepositAmount: undefined,
        OverriddenDepositWithoutCarryoutAmount: undefined,
        PeriodicDiscountsCalculateScopeValue: 0,
        PrepaymentAmountInvoiced: 0,
        PrepaymentAmountPaid: 0,
        PrepaymentAppliedOnPickup: 0,
        PromotionLines: [],
        QuotationExpiryDate: undefined,
        ReasonCodeLines: [],
        ReceiptEmail: 'email',
        ReceiptTransactionTypeValue: 1,
        RefundableTenderLines: [],
        RequestedDeliveryDate: undefined,
        RequiredDepositAmount: 0,
        RequiredDepositWithoutCarryoutAmount: 0,
        ReturnTransactionHasLoyaltyPayment: false,
        SalesId: '0',
        ShippingAddress: undefined,
        ShippingChargeAmount: 0,
        StaffId: '0',
        SubtotalAmount: 1580,
        SubtotalAmountWithoutTax: 1580,
        SubtotalSalesAmount: 1580,
        SuspendedCartId: '1',
        TaxAmount: 0,
        TaxOnCancellationCharge: 0,
        TaxOverrideCode: '1',
        TaxViewLines: [],
        TenderLines: [],
        TerminalId: '',
        TotalAmount: 1580,
        TotalCarryoutSalesAmount: 1580,
        TotalCustomerOrderSalesAmount: 0,
        TotalItems: 2,
        TotalManualDiscountAmount: 0,
        TotalManualDiscountPercentage: 0,
        TotalReturnAmount: 0,
        TotalSalesAmount: 1580,
        TransactionTypeValue: -1,
        Version: 22223736,
        WarehouseId: 'DC-CENTRAL'
    };
    it('renders correctly with change event', () => {
        // @ts-expect-error partial mock
        const cart: ICartState = {
            totalItemsInCart: 2,
            isProductAddedToCart: true,
            isEmpty: false,
            hasInvoiceLine: false,
            cart: subcart
        };
        const mockProps = {
            ...moduleProps,
            className: 'CartLineItem',
            cartLabel: 'Shopping bag',
            cartQtyLabel: '0',
            telemetryContent,
            id: 'data',
            typeName: 'data',
            data: {
                cart
            }
        };
        const component = render(<CartIcon {...mockProps} />);
        expect(component).toMatchSnapshot();
    });
    it('renders correctly with change event1', () => {
        // @ts-expect-error partial mock
        const cart: ICartState = {
            totalItemsInCart: 2,
            isProductAddedToCart: true,
            isEmpty: false,
            hasInvoiceLine: true
        };
        const mockProps = {
            ...moduleProps,
            className: 'CartLineItem',
            cartLabel: 'Shopping bag',
            cartQtyLabel: '0',
            telemetryContent,
            id: 'data',
            typeName: 'data',
            data: {
                cart
            }
        };
        const component = render(<CartIcon {...mockProps} />);
        expect(component).toMatchSnapshot();
    });
    it('renders correctly with change event2', () => {
        const mockProps = {
            ...moduleProps,
            className: 'CartLineItem',
            cartLabel: 'Shopping bag',
            cartQtyLabel: '0',
            telemetryContent,
            id: 'data',
            typeName: 'data',
            data: {
                cart: undefined
            }
        };
        const component = render(<CartIcon {...mockProps} />);
        expect(component).toMatchSnapshot();
    });
    it('renders correctly with change event3', () => {
        // @ts-expect-error partial mock
        const cart: ICartState = {
            totalItemsInCart: 120,
            isProductAddedToCart: true,
            isEmpty: false,
            hasInvoiceLine: false,
            cart: subcart
        };
        const mockProps = {
            ...moduleProps,
            className: 'CartLineItem',
            cartLabel: 'Shopping bag',
            cartQtyLabel: '0',
            telemetryContent,
            id: 'data',
            typeName: 'data',
            data: {
                cart
            }
        };
        const component = render(<CartIcon {...mockProps} />);
        expect(component).toMatchSnapshot();
    });
});
