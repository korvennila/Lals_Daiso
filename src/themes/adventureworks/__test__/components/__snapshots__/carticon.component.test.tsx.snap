// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartIcon renders correctly with change event 1`] = `
initialize {
  "0": Node {
    "attribs": Object {
      "aria-label": "Shopping bag",
      "class": "msc-cart-icon",
      "data-bi-id": "12345",
      "data-m": "{\\"id\\":\\"12345\\",\\"key1\\":\\"val1\\"}\\"",
      "role": "button",
      "title": "Shopping bag",
    },
    "children": Array [
      Node {
        "attribs": Object {
          "aria-hidden": "true",
          "class": "msc-cart-icon__count",
          "style": "visibility:visible",
        },
        "children": Array [
          Node {
            "data": "0",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "aria-hidden": undefined,
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "aria-hidden": undefined,
          "class": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "aria-label": undefined,
      "class": undefined,
      "data-bi-id": undefined,
      "data-m": undefined,
      "role": undefined,
      "title": undefined,
    },
    "x-attribsPrefix": Object {
      "aria-label": undefined,
      "class": undefined,
      "data-bi-id": undefined,
      "data-m": undefined,
      "role": undefined,
      "title": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`CartIcon renders correctly with change event1 1`] = `
initialize {
  "0": Node {
    "attribs": Object {
      "aria-label": "Shopping bag",
      "class": "msc-cart-icon",
      "data-bi-id": "12345",
      "data-m": "{\\"id\\":\\"12345\\",\\"key1\\":\\"val1\\"}\\"",
      "role": "button",
      "title": "Shopping bag",
    },
    "children": Array [
      Node {
        "attribs": Object {
          "aria-hidden": "true",
          "class": "msc-cart-icon__count",
          "style": "visibility:visible",
        },
        "children": Array [
          Node {
            "data": "0",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "aria-hidden": undefined,
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "aria-hidden": undefined,
          "class": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "aria-label": undefined,
      "class": undefined,
      "data-bi-id": undefined,
      "data-m": undefined,
      "role": undefined,
      "title": undefined,
    },
    "x-attribsPrefix": Object {
      "aria-label": undefined,
      "class": undefined,
      "data-bi-id": undefined,
      "data-m": undefined,
      "role": undefined,
      "title": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`CartIcon renders correctly with change event2 1`] = `
initialize {
  "0": Node {
    "attribs": Object {
      "aria-label": "Shopping bag",
      "class": "msc-cart-icon",
      "data-bi-id": "12345",
      "data-m": "{\\"id\\":\\"12345\\",\\"key1\\":\\"val1\\"}\\"",
      "role": "button",
      "title": "Shopping bag",
    },
    "children": Array [
      Node {
        "attribs": Object {
          "aria-hidden": "true",
          "class": "msc-cart-icon__count",
          "style": "visibility:hidden",
        },
        "children": Array [
          Node {
            "data": "0",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "aria-hidden": undefined,
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "aria-hidden": undefined,
          "class": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "aria-label": undefined,
      "class": undefined,
      "data-bi-id": undefined,
      "data-m": undefined,
      "role": undefined,
      "title": undefined,
    },
    "x-attribsPrefix": Object {
      "aria-label": undefined,
      "class": undefined,
      "data-bi-id": undefined,
      "data-m": undefined,
      "role": undefined,
      "title": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`CartIcon renders correctly with change event3 1`] = `
initialize {
  "0": Node {
    "attribs": Object {
      "aria-label": "Shopping bag",
      "class": "msc-cart-icon",
      "data-bi-id": "12345",
      "data-m": "{\\"id\\":\\"12345\\",\\"key1\\":\\"val1\\"}\\"",
      "role": "button",
      "title": "Shopping bag",
    },
    "children": Array [
      Node {
        "attribs": Object {
          "aria-hidden": "true",
          "class": "msc-cart-icon__count",
          "style": "visibility:visible",
        },
        "children": Array [
          Node {
            "data": "99+",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "aria-hidden": undefined,
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "aria-hidden": undefined,
          "class": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "aria-label": undefined,
      "class": undefined,
      "data-bi-id": undefined,
      "data-m": undefined,
      "role": undefined,
      "title": undefined,
    },
    "x-attribsPrefix": Object {
      "aria-label": undefined,
      "class": undefined,
      "data-bi-id": undefined,
      "data-m": undefined,
      "role": undefined,
      "title": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
